reconcile/vault_replication.py

```diff
@@ -0,0 +1,225 @@
+            and policy.instance.name == instance_name
+        ):
+            for line in policy.rules.split("\n"):
+                res = re.search(r"path \s*[\'\"](.*)[\'\"]", line)
+
+                if res is not None:
+                    policy_paths.append(res.group(1))
+
+    return policy_paths
+
+
+def get_jenkins_secret_list(jenkins_instance: str) -> List[str]:
+    secret_list = []
+    query_data = jenkins_configs.query(query_func=gql.get_api().query)
+
+    for p in query_data.jenkins_configs:
+        if (
+            isinstance(p, JenkinsConfigV1)
+            and p.instance.name == jenkins_instance
+            and p.config_path
+        ):
+            secret_paths = [
+                line
+                for line in p.config_path.content.split("\n")
+                if "secret-path" in line
+            ]
+            for line in secret_paths:
+                res = re.search(r"secret-path:\s*[\'\"](.*)[\'\"]", line).group(1)
+                secret_list.append(res)
+
+    return secret_list
+
+
+def get_vault_credentials(vault_instance: vault_instances.VaultInstanceV1):
+    vault_creds = {"server": "", "role_id": None, "secret_id": None}
+    vault = cast(_VaultClient, VaultClient())
+
+    if isinstance(vault_instance, VaultReplicationV1):
+        vault_instance = vault_instance.instance
+
+    role_id = {
+        "path": vault_instance.auth.role_id.path,
+        "field": vault_instance.auth.role_id.field,
+    }
+    secret_id = {
+        "path": vault_instance.auth.secret_id.path,
+        "field": vault_instance.auth.secret_id.field,
+    }
+
+    vault_creds["role_id"] = vault.read(role_id)
+    vault_creds["secret_id"] = vault.read(secret_id)
+    vault_creds["server"] = vault_instance.address
+
+    return vault_creds
+
+
+def replicate_paths(
+    dry_run: bool, source_vault: _VaultClient, dest_vault: _VaultClient, replications
+) -> None:
+    for path in replications.paths:
+
+        provider = path.provider
+
+        if provider == "jenkins":
+            if path.policy is not None:
+                policy_paths = get_policy_paths(
+                    path.policy.name,
+                    path.policy.instance.name,
+                )
+            else:
+                policy_paths = None
+
+            path_list = get_jenkins_secret_list(path.jenkins_instance.name)
+            check_copy_secret_list(
+                dry_run, source_vault, dest_vault, path_list, policy_paths
+            )
+
+
+def run(dry_run: bool) -> None:
+
+    query_data = vault_instances.query(query_func=gql.get_api().query)
+
+    for instance in query_data.vault_instances:
+        if isinstance(instance, VaultInstanceV1) and instance.replication:
+            for replication in instance.replication:
+                source_creds = get_vault_credentials(instance)
+                dest_creds = get_vault_credentials(replication.vault_instance)
+
+                source_vault = cast(
+                    _VaultClient,
```

Comment:
I don't understand this very well. Why do you cast instead of using directly `_VaultClient`? If I understand correctly, you need to do it because `VaultClient` is a singleton.

Reply:
Fixed the function to use _VaultClient, didn't noticed that was a singleton, but make sense for the previous use cases

Reply:
great. Please add a comment explaining why you need to do it.